File: java/ultimate/AbstractMiddlewareHandler.java
package ultimate;

public class AbstractMiddlewareHandler implements MiddlewareHandler {

    private MiddlewareHandler next;

    @Override
    public MiddlewareHandler setNext(MiddlewareHandler middlewareHandler) {
        next = middlewareHandler;
        return next;
    }

    @Override
    public MiddlewareResponse handle(HttpRequest request) {
        if (next != null) {
            next.handle(request);
        }
        return new MiddlewareResponse("Request successfully passed all middlewares", true);
    }
}

File: typescript/patterns/ultimate-design-pattern-challenge/abstract-middleware-handler.ts
class AbstractMiddlewareHandler implements MiddlewareHandler {
  private next: MiddlewareHandler | null = null;

  setNext(middlewareHandler: MiddlewareHandler): MiddlewareHandler {
    this.next = middlewareHandler;
    return this.next;
  }

  handle(request: HttpRequest): MiddlewareResponse {
    if (!this.next) {
      return new MiddlewareResponse(
        "Request successfully passed all middlewares",
        true,
      );
    }

    return this.next.handle(request);
  }
}
